File: LeetCode/Easy/0637-average-of-levels-in-binary-tree/0637-average-of-levels-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    static class Node {
        TreeNode treenode;
        int depth;
        Node (TreeNode treenode, int depth) {
            this.treenode = treenode;
            this.depth = depth;
        }
    }
    public List<Double> averageOfLevels(TreeNode root) {
        ArrayDeque<Node> q = new ArrayDeque<>();
        List<Double> ans = new ArrayList<>();
        double[] sumOfLevels = new double[10000];
        int[] countOfLevels = new int[10000];
        int maxDepth = 0;

        q.add(new Node(root, 0));
        sumOfLevels[0] = root.val;
        countOfLevels[0] = 1;

        
        while (!q.isEmpty()) {
            Node cur = q.removeFirst();
            if (cur.treenode.left != null) {
                q.add(new Node(cur.treenode.left, cur.depth + 1));
                sumOfLevels[cur.depth + 1] += cur.treenode.left.val;
                countOfLevels[cur.depth + 1] += 1;
                maxDepth = Math.max(maxDepth, cur.depth + 1);
            }
            if (cur.treenode.right != null) {
                q.add(new Node(cur.treenode.right, cur.depth + 1));
                sumOfLevels[cur.depth + 1] += cur.treenode.right.val;
                countOfLevels[cur.depth + 1] += 1;
                maxDepth = Math.max(maxDepth, cur.depth + 1);
            }
        }

        for (int i = 0; i <= maxDepth; i++) {
            ans.add(Double.valueOf(sumOfLevels[i]/countOfLevels[i]));
        }

        return ans;
    }
}

File: LeetCode/Easy/0637-average-of-levels-in-binary-tree/0637-average-of-levels-in-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

 type QueueNode struct {
    Node *TreeNode
    Depth int
 }

func averageOfLevels(root *TreeNode) []float64 {
    n := depthOfTreeNode(root)
    avgArr := make([]float64, n)
    cntArr := make([]int, n)

    q := make([]QueueNode, 0)

    q = append(q, QueueNode{root, 1})
    avgArr[0] = float64(root.Val)
    cntArr[0] = 1

    for (len(q) != 0) {
        cur := q[0]
        q = q[1:]

        sum := 0
        cnt := 0

        if (cur.Node.Left != nil) {
            sum += cur.Node.Left.Val
            cnt += 1
            q = append(q, QueueNode{cur.Node.Left, cur.Depth + 1})
        }

        if (cur.Node.Right != nil) {
            sum += cur.Node.Right.Val
            cnt += 1
            q = append(q, QueueNode{cur.Node.Right, cur.Depth + 1})
        }

        if (cnt >= 1){
            avgArr[cur.Depth] += float64(sum)
            cntArr[cur.Depth] += cnt
        }
    }

    for i := 0; i < n; i++ {
        avgArr[i] = avgArr[i] / float64(cntArr[i])
    }


    return avgArr
}

func depthOfTreeNode(root *TreeNode) int {
    if (root == nil) {
        return 0
    }
    return max(depthOfTreeNode(root.Left), depthOfTreeNode(root.Right)) + 1
}